Add EIP-5453: Smart Contract Crypto Endorsement (#5453)
ethereum/EIPs: EIPS/eip-5453.md

        string memory erc1271Prefix = "\x19Ethereum Signed Message:\n32";
        bytes32 erc1271Hash = keccak256(abi.encodePacked(erc1271Prefix, _msgDigest));
        address endorser = ecrecover(erc1271Hash,
            uint8(_end[128]), // v
            bytes32(_end[64:96]), // r
            bytes32(_eSig[96:128]) // s
        );
        // _isEligibleEndorser is application specific.
        require (_isEligibleEndorser(endorser), "Endorser is not eligible to transfer.");
        _;
    }
}

contract EndorsableERC721 is SomeERC721, ERC5453 {
  function safeTransferFrom(
    address from,
    address to,
    uint256 id,
    bytes calldata extraData)

    onlyEndorsed(  // used as modifier
        keccak256(
            abi.encodePacked(
                from, to, id, amount,
                extraData[:REMAIN_LENGTH], // first part of extraData is reserved for original use for extraData unendorsed.
                extraData[REMAIN_LENGTH: REMAIN_LENGTH + 32], // nonce of endorsement for the {contract, endorser} combination
                extraData[REMAIN_LENGTH + 32: REMAIN_LENGTH + 64], // valid_by for the endorsement
            ),

>         data[REMAIN_LENGTH:] // the endorsement component

Reviewer: Shouldn't it be `extraData[REMAIN_LENGTH:]`?

Author: It might. There is issue in this rrefimpl. Lets talk later today soon

  3. For `Erc5453FormatType == FIRST_TYPE_SINGLE_ENDORSER`, it supports a simple endorsement signature in the following format. Any complying contract MUST implement FIRST_TYPE. Other type number will be used for future extension, such as a multi-but-separate-endorsement scenario.
  4. Complying contract MUST emit the following `OnEndorsed` event when a transaction is executed because of the endorsement. Complying contract MUST NOT emit such event when a transaction is executed without taking account of the endorsement.

```solidity
event OnEndorsed(byte32 MethodABIKeccak, address[] memory endorsers);
```

4. The format of `FIRST_TYPE_SINGLE_ENDORSER` type is

```solidity
endorser_nonce(uint256, 32bytes) ||
valid_by(uint256, 32bytes) ||
r,s,v (32bytes + 32bytes + 1bytes = 65bytes)||
MAGIC_WORLD(8bytes)
```

The total length denoted as `LENGTH_OF_ENDORSEMENT = 137 bytes`

5. The `(r,s,v)` are an ECDSA signature of all method fields plus the head remainder of `extraData` removing the last `LENGTH_OF_ENDORSEMENT` bytes. See the [Reference Implementation](#reference-implementation) for clarification.

## Rationale

> 1. Choosing [EIP-5269](./eip-5269.md) instead of [EIP-165](./eip-165.md) because the data format cannot be represented by method siganture as in [EIP-165](./eip-165.md).

Reviewer 2: ```suggestion
1. Choosing [EIP-5269](./eip-5269.md) instead of [EIP-165](./eip-165.md) because the data format cannot be represented by method signature as in [EIP-165](./eip-165.md).
```

---
eip: 5453
title: Smart Contract Crypto Endorsement
description: A data format for including digital signatures in a smart contract function call.
author: Zainan Victor Zhou (@xinbenlv)
discussions-to: https://ethereum-magicians.org/t/erc-5453-endorsement-standard/10355
status: Draft
type: Standards Track
category: ERC
created: 2022-08-12

> requires: 5269

Reviewer 2: ```suggestion
requires: 20, 165, 721, 1155, 5269
```

if the ending 8-bytes doesn't matches with this magic word.
  2. the last 2-bytes MUST be a type-indicator, denoted as `Erc5453FormatType`
  3. For `Erc5453FormatType == FIRST_TYPE_SINGLE_ENDORSER`, it supports a simple endorsement signature in the following format. Any complying contract MUST implement FIRST_TYPE. Other type number will be used for future extension, such as a multi-but-separate-endorsement scenario.
  4. Complying contract MUST emit the following `OnEndorsed` event when a transaction is executed because of the endorsement. Complying contract MUST NOT emit such event when a transaction is executed without taking account of the endorsement.

```solidity
event OnEndorsed(byte32 MethodABIKeccak, address[] memory endorsers);
```

4. The format of `FIRST_TYPE_SINGLE_ENDORSER` type is

```solidity
endorser_nonce(uint256, 32bytes) ||
valid_by(uint256, 32bytes) ||
r,s,v (32bytes + 32bytes + 1bytes = 65bytes)||
MAGIC_WORLD(8bytes)
```

The total length denoted as `LENGTH_OF_ENDORSEMENT = 137 bytes`

> 5. The `(r,s,v)` are a ECDSA signature of all method fields plus the head remainder of `extraData` removing
> the last LENGTH_OF_ENDORSEMENT bytes. See reference implementation for clarification.

Reviewer 2: ```suggestion
5. The `(r,s,v)` are an ECDSA signature of all method fields plus the head remainder of `extraData` removing the last `LENGTH_OF_ENDORSEMENT` bytes. See the [Reference Implementation](#reference-implementation) for clarification.
```

Formatting change required, grammatical fix required, link not required.

## Rationale

1. Choosing [EIP-5269](./eip-5269.md) instead of [EIP-165](./eip-165.md) because the data format cannot be represented by method siganture as in [EIP-165](./eip-165.md).

1. Originally we considered adding the endorsement at the first part of `extraData`. We turn into adding the endorsement to the ending of data, which allows the following future extensions such as:

- Allow general endorsable be applied
- Allow nested endorsements

## Backwards Compatibility

> The design assumes a `bytes calldata extraData` to maximize the flexibility of future extensions.
> This assumption is compatible with [EIP-721](eip-721.md), [EIP-1155](eip-1155.md) and many other ERC-track EIPs.
> Those that aren't, such as [EIP-20](./eip-20.md), can also be updated to support it,
> such as using a wrapper contract or proxy upgrade.

Reviewer 2: ```suggestion
The design assumes a `bytes calldata extraData` to maximize the flexibility of future extensions. This assumption is compatible with [EIP-721](eip-721.md), [EIP-1155](eip-1155.md) and many other ERC-track EIPs. Those that aren't, such as [EIP-20](./eip-20.md), can also be updated to support it, such as using a wrapper contract or proxy upgrade.
```
Formatting change required.

## Reference Implementation

```solidity

> 

Reviewer 2: ```suggestion
```

        address endorser = ecrecover(erc1271Hash,
            uint8(_end[128]), // v
            bytes32(_end[64:96]), // r
            bytes32(_eSig[96:128]) // s
        );
        // _isEligibleEndorser is application specific.
        require (_isEligibleEndorser(endorser), "Endorser is not eligible to transfer.");
        _;
    }
}

contract EndorsableERC721 is SomeERC721, ERC5453 {
  function safeTransferFrom(
    address from,
    address to,
    uint256 id,
    bytes calldata extraData)

    onlyEndorsed(  // used as modifier
        keccak256(
            abi.encodePacked(
                from, to, id, amount,
                extraData[:REMAIN_LENGTH], // first part of extraData is reserved for original use for extraData unendorsed.
                extraData[REMAIN_LENGTH: REMAIN_LENGTH + 32], // nonce of endorsement for the {contract, endorser} combination
                extraData[REMAIN_LENGTH + 32: REMAIN_LENGTH + 64], // valid_by for the endorsement
            ),
        data[REMAIN_LENGTH:] // the endorsement component
    ) {
    super.safeTransferFrom(from, to, id, extraData[:REMAIN_LENGTH]); // original
  }
}
```

## Security Considerations

> ### Replay-attack

Reviewer 2: ```suggestion
### Replay Attacks
```

> Replay attack is a type of attack on cryptography authentication.
> In a narrower sense, it usually refers to is a type of attack that circumvent the crypto
> signature verification by reusing an existing signature for message being signed again.
> Any implementations relying on this ERC must realize that all smart endorsements
> described here are cryptography signatures and are *public* and can be obtained by anyone.
> They must forsee the possibility of replay of the transactions not only
> at the exactly deployment of the same smart contract, but also other deployment
> of similar smart contract source, or a version on another chainId etc.
> All the `nonce`, `valid_since`, `valid_by` were meant to restrict the surface of attack but
> might not eliminate risks of all such attacks, e.g. see **Phishing** for endorsement.

Reviewer 2: ```suggestion
A replay attack is a type of attack on cryptography authentication. In a narrow sense, it usually refers to a type of attack that circumvents the cryptographically signature verification by reusing an existing signature for a message being signed again. Any implementations relying on this EIP must realize that all smart endorsements described here are cryptographic signatures that are *public* and can be obtained by anyone. They must foresee the possibility of a replay of the transactions not only at the exact deployment of the same smart contract, but also other deployments of similar smart contracts, or of a version of the same contract on another `chainId`, or any other similar attack surfaces. The `nonce`, `valid_since`, and `valid_by` fields are meant to restrict the surface of attack but might not fully eliminate the risk of all such attacks, e.g. see the [Phishing](#phishing) section.
```
The wording change is optional (but recommended since it fixes the grammatical errors), but the formatting change is required.

### Phishing

> It's worth pick out a special form of replay-attack by phishing.
> An adversary can design another smart contract in the way that user be tricked
> into signing a smart endorsement for a seemingly legitimate purpose,
> but the data-to-designed matches the target application

Reviewer 2: ```suggestion
It's worth pointing out a special form of replay attack by phishing. An adversary can design another smart contract in a way that the user be tricked into signing a smart endorsement for a seemingly legitimate purpose, but the data-to-designed matches the target application
```

The wording change is optional (but recommended since it fixes the grammatical errors), but the formatting change is required.

> 

Reviewer 2: ```suggestion
```

---
eip: 5453
title: Smart Endorsement
description: Defines a format for endorsement on-chain
author: Zainan Victor Zhou (@xinbenlv)
discussions-to: https://ethereum-magicians.org/t/erc-5453-endorsement-standard/10355
status: Draft
type: Standards Track

> category: Interface

Reviewer 2: ```suggestion
category: ERC
```

created: 2022-08-12
requires: 165
---

> <!-- TODO add 5269 in requires after the ./eip-5269.md becomes available. -->

Reviewer 2: Just a warning: EIP-5269 will block EIP-5453 if that is the case.

## Abstract

Provides a format for endorsement of smart contract transactions in the format of extra data (the last bytes field of a method signature).

## Motivation

1. Support a second approval from another user.
2. Support pay-for-by another user
3. Support multi-sig
4. Support persons acting in concert by endorsements
5. Support accumulated voting
6. Support off-line signatures

> 
> TODO add implementation references for each of the case above.

Reviewer 2: ```suggestion
```

> ## Specification

Reviewer 2: ```suggestion
## Specification

```

Author: Committed

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

1. A field of method signature MUST be designated to be the endorsement field.
It must exceed a minimum length of `65 bytes` which is the length of (v, r, s) where as

> 

Reviewer 2: ```suggestion
```

- `r, s` are the ECDSA recover signature pair. `v`'s last bit carries ECDSA recover Y-Parity.

> 

Reviewer 2: ```suggestion
```

2. For any complying method of compliant contract, the last data field MUST be `bytes`.

> 

Reviewer 2: ```suggestion
```

3. Such data field MUST conform to the following format

> 
> 3.1. the last 8-bytes MUST be a magic 8-bytes word in the ASCII representation of "ENDORSED".
> Implementing method of smart contract that honors smart endorsement MUST deemed a transaction unendorsed by anyone,
> if the ending 8-bytes doesn't matches with this magic word.
> 3.2. the last 2-bytes MUST be a type-indicator, denoted as `Erc5453FormatType`
> 3.3. For `Erc5453FormatType == FIRST_TYPE_SINGLE_ENDORSER` <!--TODO determine a type number-->,
> it supports a simple endorsement signature in the following format. Any complying contract MUST implement FIRST_TYPE.
> Other type number will be used for future extension, such as a multi-but-separate-endorsement scenario.
> 
> 3.4. Complying contract MUST emit the following `OnEndorsed` event when a transaction is executed because of the endorsement. Complying contract MUST NOT emit such event when a transaction is executed without taking account of the endorsement.

Reviewer 2: ```suggestion
  1. the last 8-bytes MUST be a magic 8-bytes word in the ASCII representation of "ENDORSED". Implementing method of smart contract that honors smart endorsement MUST deemed a transaction unendorsed by anyone,
if the ending 8-bytes doesn't matches with this magic word.
  2. the last 2-bytes MUST be a type-indicator, denoted as `Erc5453FormatType`
  3. For `Erc5453FormatType == FIRST_TYPE_SINGLE_ENDORSER`, it supports a simple endorsement signature in the following format. Any complying contract MUST implement FIRST_TYPE. Other type number will be used for future extension, such as a multi-but-separate-endorsement scenario.
  4. Complying contract MUST emit the following `OnEndorsed` event when a transaction is executed because of the endorsement. Complying contract MUST NOT emit such event when a transaction is executed without taking account of the endorsement.
```

```solidity
event OnEndorsed(byte32 MethodABIKeccak, address[] memory endorsers);
```

4. The format of `FIRST_TYPE_SINGLE_ENDORSER` type is

```solidity
endorser_nonce(uint256, 32bytes) ||
valid_by(uint256, 32bytes) ||
r,s,v (32bytes + 32bytes + 1bytes = 65bytes)||
MAGIC_WORLD(8bytes)
```

The total length denoted as `LENGTH_OF_ENDORSEMENT = 137 bytes`

5. The `(r,s,v)` are a ECDSA signature of all method fields plus the head remainder of `extraData` removing
the last LENGTH_OF_ENDORSEMENT bytes. See reference implementation for clarification.

## Rationale

<!-- TODO add the following after EIP-5269 becomes available

> 1. Choosing [EIP-5269] instead of [EIP-165](./eip-165.md) because the data format cannot be represented by format.

Reviewer 2: I suggest making a minimal solidity interface so that you *can* use EIP-165.

```solidity
function supports5453() pure returns (bool)
```

Author: That's not the purpose of this design. (Also one of the case that EIP-165 is not suitable for identifying all interfaces)

Reviewer 2: This isn't an interface. I'd say it's more of a design choice.

Author: I lean towards to *not* expose `supports5453()` because it's unnecessarily and unclear - Largely because whether a contract supports endorsement behavior will depend on caller - user who are eligible to sign endorsement. Therefore it would be not very meaningful to expose it in EIP-165.

 -->

1. Originally we considered adding the endorsement at the first part of `extraData`. We turn into adding the endorsement to the ending of data, which allows the following future extensions such as:

> - Allow nested endorsements for nested function calls. TODO add reference implementation

Reviewer 2: ```suggestion
```

- Allow general endorsable be applied

## Backwards Compatibility

The design assumes a `bytes calldata extraData` to maximize the flexibility of future extensions.
This assumption is compatible with [EIP-721](eip-721.md), [EIP-1155](eip-1155.md) and many other ERC-track EIPs.
Those that aren't, such as [EIP-20](./eip-20.md), can also be updated to support it,
such as using a wrapper contract or proxy upgrade.

## Test Cases
TODO

> ## Reference Implementation

Reviewer 2: ```suggestion
## Reference Implementation

```

                extraData[REMAIN_LENGTH + 32: REMAIN_LENGTH + 64], // valid_by for the endorsement
            ),
        data[REMAIN_LENGTH:] // the endorsement component
    ) {
    super.safeTransferFrom(from, to, id, extraData[:REMAIN_LENGTH]); // original
  }
}
```

## Security Considerations

- `Replay-attack`: Replay attack is a type of attack on cryptography authentication.
In a narrower sense, it usually refers to is a type of attack that circumvent the crypto
signature verification by reusing an existing signature for message being signed again.
Any implementations relying on this ERC must realize that all smart endorsements
described here are cryptography signatures and are *public* and can be obtained by anyone.
They must forsee the possibility of replay of the transactions not only
at the exactly deployment of the same smart contract, but also other deployment
of similar smart contract source, or a version on another chainId etc.
All the `nonce`, `valid_since`, `valid_by` were meant to restrict the surface of attack but
might not eliminate risks of all such attacks, e.g. see **Phishing** for endorsement.

> - `Phishing`: It's worth pick out a special form of replay-attack by phishing.

Reviewer 2: "Replay attacks" isn't hyphenated. Likewise, you're splitting the lines manually.

Author: Updated accordingly

An adversary can design another smart contract in the way that user be tricked
into signing a smart endorsement for a seemingly legitimate purpose,
but the data-to-designed matches the target application

> - `nonce` is required to avoid replay attack.

Reviewer 2: ```suggestion
- A `nonce` is required to avoid replay attacks.
```

Author: Removed mentioning of nonce and valid_by

- `valid_by` is needed to ensure a confirmed end of life-cycle for an endorsement.

> ### Related EIPs
> 
> - [EIP-712](./eip-712.md)
> - [EIP-2242](./eip-2242.md) proposes postdata: proposed to be unreadable to EVM. It's possible new EIP
> proposes supporting a different type of `postData` via Solidity compilation or general additional data
> field.

Reviewer 2: ```suggestion
```

---
eip: 5453
title: Smart Endorsement
description: Defines a format for endorsement on-chain
author: Zainan Victor Zhou (@xinbenlv)
discussions-to: https://ethereum-magicians.org/t/erc-5453-endorsement-standard/10355
status: Draft
type: Standards Track
category: Interface
created: 2022-08-12

> requires: 5269

Reviewer 2: This EIP Doesn't exist.

Author: It now does

---

## Abstract

Provides a format for endorsement of smart contract transactions in the format of extra data (the last bytes field of a method signature).

## Motivation

1. Support a second approval from another user: TODO
2. Support pay-for-by: TODo
3. Support multi-sig: TODO
4. Support persons acting in concert by endorsements: TODO
5. Support accumulated voting: TODO
6. Support off-line signatures: TODO

## Specification
The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

```solidity

> 

Reviewer 2: Where's the code?

Author: Added the code

```

## Rationale

> 1. Choosing [EIP-5269](./eip-5269.md) instead of [EIP-165](./eip-165.md) because the data format cannot be represneted by format.

Reviewer 2: ```suggestion
1. Choosing [EIP-5269](./eip-5269.md) instead of [EIP-165](./eip-165.md) because the data format cannot be represented by format.
```
This is unclear. You're also referencing something that doesn't exist.

Author: It now does

Reviewer 2: The typo still needs fixing though.

## Backwards Compatibility

> The design is assumes a `bytes calldata extraData` and hence to maximize the flexibility of future extension. It is compatible with popular ERCs such as ERC-721, ERC-1155. ERC-20 can also be updated to support it, such as wrapper contract or proxy upgrade.

Reviewer 2: ```suggestion
The design assumes a `bytes calldata extraData` to maximize the flexibility of future extensions. It is compatible with most popular EIPs, and those that aren't can also be updated to support it, such as wrapper contract or proxy upgrade.
```

Author: You were so fast. it now contains code.

> ## Test Cases
> TODO
> 
> ## Reference Implementation
> TODO

Reviewer 2: ```suggestion
```
